File: api/src/modules/notifications/push.service.ts
import { Injectable, Logger } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import { DevicePushTokenEntity } from '../user/entities/device-push-token.entity';

@Injectable()
export class PushService {
  private readonly logger = new Logger(PushService.name);

  constructor(
    @InjectRepository(DevicePushTokenEntity)
    private readonly pushTokenRepo: Repository<DevicePushTokenEntity>,
  ) {}

  async sendExpoPushNotification(
    userId: string,
    title: string,
    body: string,
    url?: string,
  ): Promise<void> {
    try {
      const tokens = await this.pushTokenRepo.find({
        where: { user: { id: userId } },
      });

      if (tokens.length === 0) {
        return;
      }

      const expoTokens = tokens.map((t) => t.token);

      const message: any = {
        to: expoTokens,
        title,
        body,
        sound: 'default',
      };

      if (url) {
        message.data = { url };
      }

      const response = await fetch('https://exp.host/--/api/v2/push/send', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(message),
      });

      if (!response.ok) {
        const errorText = await response.text();
        this.logger.error(
          `Failed to send expo push notification: ${response.status} ${errorText}`,
        );
      }
    } catch (error) {
      this.logger.error(`Error sending expo push notification: ${error}`);
    }
  }
}

File: api-rust/src/modules/notifications/push_service.rs
use crate::app_state::AppState;
use crate::common::ApiResult;
use reqwest::Client;
use sea_orm::{ColumnTrait, EntityTrait, QueryFilter};
use serde_json::json;
use uuid::Uuid;

pub async fn send_expo_push_notification(
    state: &AppState,
    user_id: Uuid,
    title: &str,
    body: &str,
    url: Option<&str>,
) -> ApiResult<()> {
    // 1. Get push tokens for the user
    use crate::modules::users::device_push_token_entity::{Column as PushTokenColumn, Entity as PushTokenEntity};

    let tokens = PushTokenEntity::find()
        .filter(PushTokenColumn::UserId.eq(user_id))
        .all(&state.db)
        .await?;

    if tokens.is_empty() {
        return Ok(());
    }

    let expo_tokens: Vec<String> = tokens.into_iter().map(|t| t.token).collect();

    // 2. Prepare payload
    let mut message = json!({
        "to": expo_tokens,
        "title": title,
        "body": body,
        "sound": "default",
    });

    if let Some(u) = url {
        message["data"] = json!({ "url": u });
    }

    // 3. Send via reqwest
    let client = Client::new();
    let res = client
        .post("https://exp.host/--/api/v2/push/send")
        .json(&message)
        .send()
        .await;

    if let Err(e) = res {
        tracing::error!("Failed to send expo push notification: {}", e);
    }

    Ok(())
}
